File: src/main/java/org/b0n541/pmcts/mcts/TreeNodeStatistics.java
package org.b0n541.pmcts.mcts;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TreeNodeStatistics {
    private final Map<String, Double> totalScores = new HashMap<>();
    private long visits;

    public TreeNodeStatistics(final List<String> players) {
        if (players.size() == 0) {
            throw new IllegalArgumentException("At least one player needs to be defined.");
        }
        for (final String player : players) {
            totalScores.put(player, 0.0);
        }
    }

    public void addScores(final Map<String, Double> scores) {
        for (final String player : scores.keySet()) {
            checkPlayer(player);
        }

        visits++;
        scores.forEach((key, value) -> totalScores.put(key, totalScores.get(key) + value));
    }

    public long getVisits() {
        return visits;
    }

    public double getTotalScore(final String player) {
        checkPlayer(player);

        return totalScores.get(player);
    }

    private void checkPlayer(final String player) {
        if (!totalScores.containsKey(player)) {
            throw new IllegalArgumentException("Unknown player " + player);
        }
    }

    public Map<String, Double> getTotalScores() {
        return Collections.unmodifiableMap(totalScores);
    }
}

File: src/main/kotlin/net/b0n541/ai/algorithm/mcts/tree/TreeNodeStatistics.kt
package net.b0n541.ai.algorithm.mcts.tree

class TreeNodeStatistics(players: List<String>) {
    private val totalScores: MutableMap<String, Double> = mutableMapOf()
    fun totalScores() = totalScores.toMap()

    init {
        require(players.isNotEmpty()) { "At least one player needs to be defined." }

        for (player in players) {
            totalScores[player] = 0.0
        }
    }

    var visits: Long = 0
        private set

    var currentTraversals: Long = 0
        private set

    fun addTraversal() {
        //currentTraversals++
    }

    fun addScores(scores: Map<String, Double>) {
        for (player in scores.keys) {
            checkPlayer(player)
        }

        visits++
        //currentTraversals--
        scores.forEach { (key: String, value: Double) -> totalScores[key] = totalScores[key]!! + value }
    }

    fun getTotalScore(player: String?): Double {
        checkPlayer(player)
        return totalScores[player]!!
    }

    private fun checkPlayer(player: String?) {
        require(totalScores.containsKey(player)) { "Unknown player $player" }
    }
}